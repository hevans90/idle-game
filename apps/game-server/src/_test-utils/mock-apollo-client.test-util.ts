import {
  ApolloClient,
  ApolloQueryResult,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { MockSubscriptionLink } from '@apollo/client/testing';

export const mockApolloClient = () => {
  const link = new MockSubscriptionLink();

  return {
    client: new ApolloClient({
      link,
      cache: new InMemoryCache(),
    }),
    link,
  };
};

export const mockQuery = <T>(
  mockClient: ApolloClient<NormalizedCacheObject>,
  data: T
) => {
  jest.spyOn(mockClient, 'query').mockImplementation(
    () =>
      new Promise<ApolloQueryResult<T>>((resolve, reject) => {
        resolve({
          data,
          loading: false,
          networkStatus: 1,
        });
      })
  );
};
