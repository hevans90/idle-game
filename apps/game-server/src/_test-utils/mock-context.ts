import { ApolloClient, NormalizedCacheObject } from '@apollo/client';
import { Auth0API } from '../datasources/auth0';
import { Context } from '../datasources/context';
import { HasuraAPI } from '../datasources/hasura-api';
import { HasuraEmpirePurchases } from '../datasources/hasura-empire-purchases';
import { HasuraEmpireResourceModifiers } from '../datasources/hasura-empire-resource-modifiers';
import { HasuraQuestProgression } from '../datasources/hasura-quest-progression';
import { MinioAPI } from '../datasources/minio';

export const mockContext = (
  mockClient: ApolloClient<NormalizedCacheObject>,
  userId: string
): Context => {
  return {
    id: userId,
    req: undefined,
    user: undefined,
    dataSources: {
      minio: new MinioAPI(null),
      hasuraAPI: new HasuraAPI(mockClient),
      hasuraEmpirePurchases: new HasuraEmpirePurchases(mockClient),
      hasuraQuestProgression: new HasuraQuestProgression(mockClient),
      hasuraEmpireResourceModifiers: new HasuraEmpireResourceModifiers(
        mockClient
      ),
      auth0API: new Auth0API(),
    },
  };
};
