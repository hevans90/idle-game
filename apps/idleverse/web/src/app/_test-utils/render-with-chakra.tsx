import { ChakraProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import React from 'react';
import { theme } from '../_theme/theme';

export const renderWithChakra = (ui) => {
  const Wrapper = ({ children }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return render(ui, { wrapper: Wrapper });
};
