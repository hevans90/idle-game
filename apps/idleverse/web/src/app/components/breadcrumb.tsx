import { useReactiveVar } from '@apollo/client';
import { Box, useColorModeValue } from '@chakra-ui/react';
import React from 'react';
import { Link } from 'react-router-dom';
import { breadCrumbsVar, galaxyConfigVar } from '../_state/reactive-variables';

export const Breadcrumb = () => {
  const bg = useColorModeValue('gray.300', 'gray.700');
  const border = useColorModeValue('gray.200', 'gray.600');

  const crumbs = useReactiveVar(breadCrumbsVar);

  const { name: galaxyName } = useReactiveVar(galaxyConfigVar);

  // Don't render a single breadcrumb.
  if (crumbs.length <= 1) {
    return null;
  }
  return (
    <Box
      padding="1rem"
      display="flex"
      flexDirection="row"
      position="absolute"
      alignItems="start"
      bgColor={bg}
      top="0"
      left="0"
      borderWidth="1px"
      borderStyle="solid"
      borderColor={border}
      borderLeft="unset"
      borderTop="unset"
      zIndex="2"
    >
      {/* Link back to any previous steps of the breadcrumb. */}
      {crumbs.map(({ name, path }, key) => {
        if (key + 1 === crumbs.length) {
          if (name === 'view-galaxy') {
            name = galaxyName;
          }

          return <span key={key}>{name}</span>;
        } else {
          return (
            <Link key={key} to={path}>
              {name}&nbsp;&gt;&nbsp;
            </Link>
          );
        }
      })}
    </Box>
  );
};
