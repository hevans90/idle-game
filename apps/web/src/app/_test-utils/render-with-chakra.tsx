import { ChakraProvider } from '@chakra-ui/react';
import { theme } from '@idleverse/theme';
import { render } from '@testing-library/react';

export const renderWithChakra = (ui) => {
  const Wrapper = ({ children }) => (
    <ChakraProvider theme={theme({ primary: 'gray', secondary: 'teal' })}>
      {children}
    </ChakraProvider>
  );

  return render(ui, { wrapper: Wrapper });
};
