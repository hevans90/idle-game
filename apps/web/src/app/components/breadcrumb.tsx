import { useReactiveVar } from '@apollo/client';
import { Box, useColorModeValue } from '@chakra-ui/react';
import { Link } from 'react-router-dom';
import useBreadcrumbs from 'use-react-router-breadcrumbs';
import { responsiveFontProps } from '../_responsive-utils/font-props';
import { planetVar } from '../_state/planet-viewer';
import { celestialVar, galaxyConfigVar } from '../_state/reactive-variables';

export const Breadcrumb = () => {
  const bg = useColorModeValue('gray.300', 'gray.700');
  const border = useColorModeValue('gray.200', 'gray.600');

  const { name: galaxyName } = useReactiveVar(galaxyConfigVar);
  const celestial = useReactiveVar(celestialVar);
  const planet = useReactiveVar(planetVar);

  const crumbs = useBreadcrumbs([
    {
      path: '/galaxies/:id',
      breadcrumb: () => <span>{galaxyName}</span>,
    },
    {
      path: '/celestials/:id',
      breadcrumb: () => <span>{celestial?.name}</span>,
    },
    {
      path: '/planets/:id',
      breadcrumb: () => <span>{planet?.name}</span>,
    },
  ]);

  // Don't render a single breadcrumb.
  if (crumbs.length <= 1) {
    return null;
  }
  return (
    <Box
      padding="1rem"
      display="flex"
      flexDirection="row"
      position="absolute"
      alignItems="start"
      bgColor={bg}
      top="0"
      left="0"
      borderWidth="1px"
      borderStyle="solid"
      borderColor={border}
      borderLeft="unset"
      borderTop="unset"
      zIndex="2"
      {...responsiveFontProps}
    >
      {crumbs.map(({ key, match, breadcrumb }, i) => (
        <span key={match.pathname}>
          <Link to={match.pathname}>
            {i !== 0 ? <>&nbsp;&gt;&nbsp;</> : null}
            {breadcrumb}
          </Link>
        </span>
      ))}
    </Box>
  );
};
