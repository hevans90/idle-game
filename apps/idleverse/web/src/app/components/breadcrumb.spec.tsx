/* eslint-disable react/jsx-no-useless-fragment */
import { renderWithChakra } from '../_test-utils/render-with-chakra';
import { Breadcrumb } from './breadcrumb';

import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

describe('breadcrumb', () => {
  it('should render', () => {
    const history = createMemoryHistory();
    history.push('/one/two/three');

    const { container } = renderWithChakra(
      <Router location={history.location} navigator={history}>
        <Breadcrumb />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
