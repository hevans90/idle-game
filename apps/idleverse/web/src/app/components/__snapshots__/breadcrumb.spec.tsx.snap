// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumb should render 1`] = `
<div>
  <div
    class="css-2el05k"
  >
    <span>
      <a
        href="/"
      >
        <span>
          Home
        </span>
      </a>
    </span>
    <span>
      <a
        href="/one"
      >
         &gt; 
        <span>
          One
        </span>
      </a>
    </span>
    <span>
      <a
        href="/one/two"
      >
         &gt; 
        <span>
          Two
        </span>
      </a>
    </span>
    <span>
      <a
        href="/one/two/three"
      >
         &gt; 
        <span>
          Three
        </span>
      </a>
    </span>
  </div>
  <span
    class="chakra-env"
    hidden=""
  />
</div>
`;
